bug: revoked iterations causes panic

### Terraform Version and Provider Version

```
$ terraform -v
Terraform v1.5.0
on linux_amd64
+ provider registry.terraform.io/hashicorp/aws v5.2.0
+ provider registry.terraform.io/hashicorp/hcp v0.60.0
```

### Affected Resource(s)

* `hcp_packer_image` data source

### Terraform Configuration Files

```hcl
provider "hcp" {}

provider "aws" {
  region = "us-east-1"
}

data "hcp_packer_image" "learn-packer-amazonlinux2" {
  bucket_name    = "learn-packer-amazonlinux2-child"
  channel        = "latest"
  cloud_provider = "aws"
  region         = "us-east-1"
}

output "hcp_packer_image" {
  value = data.hcp_packer_image.learn-packer-amazonlinux2
}
```

### Debug Output

[gist: terraform-hcp-packer-panic.txt](https://gist.githubusercontent.com/lrstanley/47e2fee0472c256128d6afbefcb63366/raw/fa8b9c5c80c67c8f7c8e9a7576dfd2cbc282e8d2/terraform-hcp-packer-panic.txt)

### Panic Output

```
data.hcp_packer_image.learn-packer-amazonlinux2: Reading...

Planning failed. Terraform encountered an error while generating this plan.

╷
│ Error: Plugin did not respond
│
│   with data.hcp_packer_image.learn-packer-amazonlinux2,
│   on main.tf line 7, in data "hcp_packer_image" "learn-packer-amazonlinux2":
│    7: data "hcp_packer_image" "learn-packer-amazonlinux2" {
│
│ The plugin encountered an error, and failed to respond to the plugin.(*GRPCProvider).ReadDataSource call. The plugin logs may contain more details.
╵

Stack trace from the terraform-provider-hcp_v0.60.0_x5 plugin:

panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x20 pc=0xdec396]

goroutine 99 [running]:
github.com/hashicorp/terraform-provider-hcp/internal/provider.dataSourcePackerImageRead({0x1207608, 0xc000136b40}, 0x0?, {0xf06480?, 0xc0005a21a0})
        github.com/hashicorp/terraform-provider-hcp/internal/provider/data_source_packer_image.go:171 +0x5d6
github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema.(*Resource).read(0xc000267c00, {0x1207640, 0xc0005171d0}, 0xd?, {0xf06480, 0xc0005a21a0})
        github.com/hashicorp/terraform-plugin-sdk/v2@v2.25.0/helper/schema/resource.go:724 +0x12e
github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema.(*Resource).ReadDataApply(0xc000267c00, {0x1207640, 0xc0005171d0}, 0xc0005e9d00, {0xf06480, 0xc0005a21a0})
        github.com/hashicorp/terraform-plugin-sdk/v2@v2.25.0/helper/schema/resource.go:943 +0x145
github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema.(*GRPCProviderServer).ReadDataSource(0xc000280db0, {0x1207640?, 0xc0005170b0?}, 0xc00028aa40)
        github.com/hashicorp/terraform-plugin-sdk/v2@v2.25.0/helper/schema/grpc_provider.go:1195 +0x38f
github.com/hashicorp/terraform-plugin-go/tfprotov5/tf5server.(*server).ReadDataSource(0xc000599f40, {0x1207640?, 0xc000516720?}, 0xc0005c9090)
        github.com/hashicorp/terraform-plugin-go@v0.14.3/tfprotov5/tf5server/server.go:658 +0x3ef
github.com/hashicorp/terraform-plugin-go/tfprotov5/internal/tfplugin5._Provider_ReadDataSource_Handler({0x1031540?, 0xc000599f40}, {0x1207640, 0xc000516720}, 0xc0001e68c0, 0x0)
        github.com/hashicorp/terraform-plugin-go@v0.14.3/tfprotov5/internal/tfplugin5/tfplugin5_grpc.pb.go:421 +0x170
google.golang.org/grpc.(*Server).processUnaryRPC(0xc0000001e0, {0x120a5e0, 0xc0005829c0}, 0xc000026d80, 0xc0005956e0, 0x19b2610, 0x0)
        google.golang.org/grpc@v1.53.0/server.go:1336 +0xd13
google.golang.org/grpc.(*Server).handleStream(0xc0000001e0, {0x120a5e0, 0xc0005829c0}, 0xc000026d80, 0x0)
        google.golang.org/grpc@v1.53.0/server.go:1704 +0xa1b
google.golang.org/grpc.(*Server).serveStreams.func1.2()
        google.golang.org/grpc@v1.53.0/server.go:965 +0x98
created by google.golang.org/grpc.(*Server).serveStreams.func1
        google.golang.org/grpc@v1.53.0/server.go:963 +0x28a

Error: The terraform-provider-hcp_v0.60.0_x5 plugin crashed!

This is always indicative of a bug within the plugin. It would be immensely
helpful if you could report the crash with the plugin's maintainers so that it
can be fixed. The output above should help diagnose the issue.
```

### Steps to Reproduce

1. Create two versions of an image in HCP Packer. Revoke both of them.
2. Reference the bucket name that contains both revoked iterations.
3. `terraform plan`

### Expected Behavior

Would expect the plan to work as expected (as nothing is actually preventing a revoked iteration from being used).

### Actual Behavior

Panics.

### Important Factoids

If I un-revoke one of the iterations, it works without issue, which makes me believe it is directly related. See also:

![image](https://github.com/hashicorp/terraform-provider-hcp/assets/1847365/a7e6865d-1f0f-49d4-acf4-85c6e2659990)

### References

* May be related: https://github.com/hashicorp/terraform-provider-hcp/issues/496
* May be related: https://github.com/hashicorp/terraform-provider-hcp/issues/493

<!--- Please keep this note for the community --->

### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

<!--- Thank you for keeping this note for the community --->

--- a/.changelog/533.txt
+++ b/.changelog/533.txt
@@ -1,0 +1,3 @@
+```release-note:bug
+Fixed panic when `hcp_packer_image` points to a channel without an assigned iteration.
+```

--- a/internal/provider/data_source_packer_image.go
+++ b/internal/provider/data_source_packer_image.go
@@ -160,7 +160,7 @@ func dataSourcePackerImageRead(ctx context.Context, d *schema.ResourceData, meta
 			channelSlug)
 		if err != nil {
 			return diag.FromErr(err)
-		} 
+		}
 		if channel.Iteration == nil {
 			return diag.Errorf("Channel does not have an assigned iteration (channel: %s)", channelSlug)
 		}

--- a/internal/provider/data_source_packer_image.go
+++ b/internal/provider/data_source_packer_image.go
@@ -160,7 +160,8 @@ func dataSourcePackerImageRead(ctx context.Context, d *schema.ResourceData, meta
 			channelSlug)
 		if err != nil {
 			return diag.FromErr(err)
-		} else if channel.Iteration == nil {
+		} 
+		if channel.Iteration == nil {
 			return diag.Errorf("Channel does not have an assigned iteration (channel: %s)", channelSlug)
 		}
 		iteration = channel.Iteration

--- a/internal/provider/data_source_packer_image.go
+++ b/internal/provider/data_source_packer_image.go
@@ -59,12 +59,13 @@ If a project is not configured in the HCP Provider config block, the oldest proj
 				Type:         schema.TypeString,
 				Optional:     true,
 				Computed:     true,
-				ExactlyOneOf: []string{"channel"},
+				ExactlyOneOf: []string{"iteration_id", "channel"},
 			},
 			"channel": {
-				Description: "The channel that points to the version of the image being retrieved. Either this or `iteration_id` must be specified. Note: will incur a billable request",
-				Type:        schema.TypeString,
-				Optional:    true,
+				Description:  "The channel that points to the version of the image being retrieved. Either this or `iteration_id` must be specified. Note: will incur a billable request",
+				Type:         schema.TypeString,
+				Optional:     true,
+				ExactlyOneOf: []string{"iteration_id", "channel"},
 			},
 			"component_type": {
 				Description: "Name of the builder that built this image. Ex: `amazon-ebs.example`.",
@@ -150,10 +151,8 @@ func dataSourcePackerImageRead(ctx context.Context, d *schema.ResourceData, meta
 		}
 	}
 
-	var channel *packermodels.HashicorpCloudPackerChannel
-
 	if channelSlug != "" {
-		channel, err = clients.GetPackerChannelBySlug(
+		channel, err := clients.GetPackerChannelBySlug(
 			ctx,
 			client,
 			loc,
@@ -161,12 +160,9 @@ func dataSourcePackerImageRead(ctx context.Context, d *schema.ResourceData, meta
 			channelSlug)
 		if err != nil {
 			return diag.FromErr(err)
+		} else if channel.Iteration == nil {
+			return diag.Errorf("Channel does not have an assigned iteration (channel: %s)", channelSlug)
 		}
-	}
-
-	// Assuming we passed the above check, the rest of the channel is not
-	// used after that,
-	if channel != nil {
 		iteration = channel.Iteration
 	}
 

--- a/internal/provider/data_source_packer_image_iteration_test.go
+++ b/internal/provider/data_source_packer_image_iteration_test.go
@@ -320,7 +320,7 @@ func upsertBuild(t *testing.T, bucketSlug, fingerprint, iterationID string) {
 	}
 }
 
-func createChannel(t *testing.T, bucketSlug, channelSlug, iterationID string) {
+func upsertChannel(t *testing.T, bucketSlug, channelSlug, iterationID string) {
 	t.Helper()
 
 	client := testAccProvider.Meta().(*clients.Client)
@@ -484,7 +484,7 @@ func TestAcc_dataSourcePacker(t *testing.T) {
 						t.Fatal(err.Error())
 					}
 					upsertBuild(t, acctestAlpineBucket, fingerprint, itID)
-					createChannel(t, acctestAlpineBucket, acctestProductionChannel, itID)
+					upsertChannel(t, acctestAlpineBucket, acctestProductionChannel, itID)
 				},
 				Config: testConfig(testAccPackerAlpineProductionImage),
 				Check: resource.ComposeTestCheckFunc(
@@ -522,7 +522,7 @@ func TestAcc_dataSourcePacker_revokedIteration(t *testing.T) {
 						t.Fatal(err.Error())
 					}
 					upsertBuild(t, acctestUbuntuBucket, fingerprint, itID)
-					createChannel(t, acctestUbuntuBucket, acctestProductionChannel, itID)
+					upsertChannel(t, acctestUbuntuBucket, acctestProductionChannel, itID)
 					// Schedule revocation to the future, otherwise we won't be able to revoke an iteration that
 					// it's assigned to a channel
 					revokeIteration(t, itID, acctestUbuntuBucket, revokeAt)

--- a/internal/provider/data_source_packer_image_test.go
+++ b/internal/provider/data_source_packer_image_test.go
@@ -139,7 +139,7 @@ func TestAcc_dataSourcePackerImage(t *testing.T) {
 						t.Fatal(err.Error())
 					}
 					upsertBuild(t, acctestImageBucket, fingerprint, itID)
-					createChannel(t, acctestImageBucket, acctestImageChannel, itID)
+					upsertChannel(t, acctestImageBucket, acctestImageChannel, itID)
 				},
 				Config: testAccPackerImageAlpineProduction,
 				Check: resource.ComposeTestCheckFunc(
@@ -183,18 +183,40 @@ func TestAcc_dataSourcePackerImage_revokedIteration(t *testing.T) {
 						t.Fatal(err.Error())
 					}
 					upsertBuild(t, acctestUbuntuImageBucket, fingerprint, itID)
-					createChannel(t, acctestUbuntuImageBucket, acctestImageChannel, itID)
+					upsertChannel(t, acctestUbuntuImageBucket, acctestImageChannel, itID)
 					// Schedule revocation to the future, otherwise we won't be able to revoke an iteration that
 					// it's assigned to a channel
 					revokeIteration(t, itID, acctestUbuntuImageBucket, revokeAt)
-					// Sleep to make sure the iteration is revoked when we test
-					time.Sleep(5 * time.Second)
 				},
 				Config: testConfig(testAccPackerImageUbuntuProduction),
 				Check: resource.ComposeTestCheckFunc(
 					resource.TestCheckResourceAttr("data.hcp_packer_image.ubuntu-foo", "revoke_at", revokeAt.String()),
 					resource.TestCheckResourceAttr("data.hcp_packer_image.ubuntu-foo", "cloud_image_id", "ami-42"),
 				),
+			},
+		},
+	})
+}
+
+func TestAcc_dataSourcePackerImage_emptyChannel(t *testing.T) {
+	resource.Test(t, resource.TestCase{
+		PreCheck:          func() { testAccPreCheck(t, map[string]bool{"aws": false, "azure": false}) },
+		ProviderFactories: providerFactories,
+		CheckDestroy: func(*terraform.State) error {
+			deleteChannel(t, acctestArchImageBucket, acctestImageChannel, true)
+			deleteBucket(t, acctestArchImageBucket, true)
+			return nil
+		},
+		Steps: []resource.TestStep{
+			{
+				PlanOnly: true,
+				PreConfig: func() {
+					upsertRegistry(t)
+					upsertBucket(t, acctestArchImageBucket)
+					upsertChannel(t, acctestArchImageBucket, acctestImageChannel, "")
+				},
+				Config:      testConfig(testAccPackerImageArchProduction),
+				ExpectError: regexp.MustCompile(`.*Channel does not have an assigned iteration.*`),
 			},
 		},
 	})
@@ -228,7 +250,7 @@ func TestAcc_dataSourcePackerImage_channelAndIterationIDReject(t *testing.T) {
 							t.Fatal(err.Error())
 						}
 						upsertBuild(t, acctestArchImageBucket, fingerprint, itID)
-						createChannel(t, acctestArchImageBucket, acctestImageChannel, itID)
+						upsertChannel(t, acctestArchImageBucket, acctestImageChannel, itID)
 					},
 					Config:      testConfig(cfg),
 					ExpectError: regexp.MustCompile("Error: Invalid combination of arguments"),
@@ -261,7 +283,7 @@ func TestAcc_dataSourcePackerImage_channelAccept(t *testing.T) {
 						t.Fatal(err.Error())
 					}
 					upsertBuild(t, acctestArchImageBucket, fingerprint, itID)
-					createChannel(t, acctestArchImageBucket, acctestImageChannel, itID)
+					upsertChannel(t, acctestArchImageBucket, acctestImageChannel, itID)
 				},
 				Config: testConfig(testAccPackerImageArchProduction),
 				Check: resource.ComposeTestCheckFunc(

--- a/internal/provider/data_source_packer_iteration_test.go
+++ b/internal/provider/data_source_packer_iteration_test.go
@@ -65,7 +65,7 @@ func TestAcc_dataSourcePackerIteration(t *testing.T) {
 						t.Fatal(err.Error())
 					}
 					upsertBuild(t, acctestIterationBucket, fingerprint, itID)
-					createChannel(t, acctestIterationBucket, acctestIterationChannel, itID)
+					upsertChannel(t, acctestIterationBucket, acctestIterationChannel, itID)
 				},
 				Config: testConfig(testAccPackerIterationAlpineProduction),
 				Check: resource.ComposeTestCheckFunc(
@@ -103,7 +103,7 @@ func TestAcc_dataSourcePackerIteration_revokedIteration(t *testing.T) {
 						t.Fatal(err.Error())
 					}
 					upsertBuild(t, acctestIterationUbuntuBucket, fingerprint, itID)
-					createChannel(t, acctestIterationUbuntuBucket, acctestIterationChannel, itID)
+					upsertChannel(t, acctestIterationUbuntuBucket, acctestIterationChannel, itID)
 					// Schedule revocation to the future, otherwise we won't be able to revoke an iteration that
 					// it's assigned to a channel
 					revokeIteration(t, itID, acctestIterationUbuntuBucket, revokeAt)

